quill-ndbc/src/main/scala/io/getquill/context/ndbc/NdbcContext.scala

```diff
@@ -37,107 +22,72 @@ abstract class NdbcContext[I <: SqlIdiom, N <: NamingStrategy, P <: PreparedStat
+    super.executeActionReturning(sql, prepare, extractor, returningBehavior)
 
-  def executeAction[T](sql: String, prepare: Prepare = identityPrepare): Future[Long] = {
-    withDataSource { ds =>
-      val (params, ps) = prepare(createPreparedStatement(sql))
-      logger.logQuery(sql, params)
+  override def executeBatchAction(groups: List[BatchGroup]): Future[List[Long]] =
+    super.executeBatchAction(groups)
 
-      ds.execute(ps).toScala.map(_.longValue)
-    }.flatMap(result => result)
-  }
+  override def executeBatchActionReturning[T](groups: List[BatchGroupReturning], extractor: Extractor[T]): Future[List[T]] =
+    super.executeBatchActionReturning(groups, extractor)
 
-  def executeActionReturning[O](sql: String, prepare: Prepare = identityPrepare, extractor: R => O, returningAction: ReturnAction): Future[O] = {
-    val expanded = expandAction(sql, returningAction)
-    executeQuerySingle(expanded, prepare, extractor)
-  }
+  override def transaction[T](f: => Future[T]): Future[T] = super.transaction(f)
 
-  def executeBatchAction(groups: List[BatchGroup]): Future[List[Long]] =
-    Future.sequence {
-      groups.map {
-        case BatchGroup(sql, prepare) =>
-          prepare.foldLeft(Future.successful(ArrayBuffer.empty[Long])) {
-            case (acc, prepare) =>
-              acc.flatMap { array =>
-                executeAction(sql, prepare).map(array :+ _)
-              }
-          }
-      }
-    }.map(_.flatten.toList)
-
-  def executeBatchActionReturning[T](groups: List[BatchGroupReturning], extractor: R => T): Future[List[T]] =
-    Future.sequence {
-      groups.map {
-        case BatchGroupReturning(sql, column, prepare) =>
-          prepare.foldLeft(Future.successful(ArrayBuffer.empty[T])) {
-            case (acc, prepare) =>
-              acc.flatMap { array =>
-                executeActionReturning(sql, prepare, extractor, column).map(array :+ _)
-              }
-          }
-      }
-    }.map(_.flatten.toList)
+  override def withDataSource[T](f: DataSource[P, R] => Future[T]): Future[T] = f(dataSource)
+  /* TODO: I'm assuming that we don't need to bracket and close the dataSource like with JDBC
+      because previously it wasn't done here either */
 
-  @tailrec
-  private def extractResult[T](rs: Iterator[R], extractor: R => T, acc: List[T] = Nil): List[T] =
-    if (rs.hasNext)
-      extractResult(rs, extractor, extractor(rs.next()) :: acc)
-    else
-      acc.reverse
+  def close(): Unit = {
+    dataSource.close()
+    ()
+  }
 
   override private[getquill] def prepareParams(statement: String, prepare: Prepare): Future[Seq[String]] =
     withDataSource { _ =>
-      prepare(createPreparedStatement(statement))._1.reverse.map(prepareParam)
+      /* TODO: I'm not sure if it's a good idea to use wrap here. The whole thing with `wrap` and
+          `withDataSource` is a mess and IMO should be cleaned up to be simpler and more compositional
+       */
```

Comment:
Would you mind changing this comment? It sounds harsh to call the work that's been done so far a mess.

Reply:
I'm going to remove all the TODO comments before merging. To be clear this wasn't meant in a bad way. I'm sure there's a reason why it was done this way but I felt I needed to point out that it feels unintuitive as a reader without the background knowledge.